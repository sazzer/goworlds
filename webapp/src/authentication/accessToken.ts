import {createReducer} from "redux-create-reducer";
import {Action, BaseAction, buildAction} from "../redux";
import produce from "immer";
import {MODULE_PREFIX} from "./module";

/** The shape of an access token */
export declare type Token = {
    token: string,
    expires: Date,
};

/** The shape of the state for this module */
export declare type State = {
    token: Token | undefined,
};

/** The initial state for this module */
const INITIAL_STATE: State = {
    token: undefined,
};

//////// Action for checking if an email address exists

/** The action type for storing an access token */
const STORE_ACCESS_TOKEN_ACTION = MODULE_PREFIX + 'storeAccessToken';

/** The shape of the action to store an access token */
declare type StoreAccessTokenAction = {
    token: string,
    expires: Date,
}

/**
 * Build the Redux action to store an access token
 * @param token The actual token
 * @param expires When the token expires
 */
export function storeAccessToken(token: string, expires: Date): Action<StoreAccessTokenAction> {
    return buildAction(STORE_ACCESS_TOKEN_ACTION, {
        token,
        expires
    });
}

/**
 * Reducer for storing the access token
 * @param state the state to update
 * @param action the action to update the state with
 * @return the new state
 */
export function storeAccessTokenReducer(state: State, action: Action<StoreAccessTokenAction>) : State {
    return produce(state, (draft: State) => {
        draft.token = action.payload;
    });
}

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [STORE_ACCESS_TOKEN_ACTION] : storeAccessTokenReducer as ((state: State, action: BaseAction) => State)
});
