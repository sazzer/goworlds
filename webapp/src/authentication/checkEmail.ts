import {from, Observable} from "rxjs";
import {request} from "../api";

/** The shape of the response from the server on checking if an email address exists */
declare type CheckEmailExistsServiceResponse = {
    exists: boolean,
};

/**
 * Check if the given Email Address is already registered
 * @param email the email address to check
 * @return the response of the check
 */
export function checkEmail(email: string) : Observable<boolean> {
    const response = request<CheckEmailExistsServiceResponse>('/emails/{email}', {
        method: 'GET',
        urlParams: {
            email,
        }
    })
        .then(response => response.body.exists);

    return from(response);
}
