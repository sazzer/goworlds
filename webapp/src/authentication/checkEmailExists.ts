import {createReducer} from "redux-create-reducer";
import produce from 'immer';
import {
    Action,
    asyncAction,
    AsyncAction,
    buildAction,
    buildSaga,
    buildSelector,
    ResolvedAsyncAction,
    startedAction,
    succeededAction
} from "../redux";
import {request} from "../api";

/** Status to indicate that we've started to check if an email address exists */
export const CHECK_STATUS_START = 'Started';

/** Status to indicate that the email address we've checked does exist */
export const CHECK_STATUS_EXISTS = 'Exists';

/** Status to indicate that the email address we've checked does not exist */
export const CHECK_STATUS_UNKNOWN = 'Unknown';

/** The shape of the state for this module */
export interface State {
    email?: string,
    state?: string,
}

/** The initial state for this module */
const INITIAL_STATE: State = {

};

/** Prefix for actions in this module */
const MODULE_PREFIX = 'CheckEmailExists/';

//////// Check If Email Exists

/** The action type for checking if an email exists */
const CHECK_EMAIL_ACTION = MODULE_PREFIX + 'checkEmailExists';

/** The action type for resetting the check for if an email exists */
const RESET_ACTION = MODULE_PREFIX + 'reset';

/** The shape of the action to check if an email exists */
type CheckEmailExistsAction = string;

/**
 * Build the Redux action to check if an email exists
 * @param email the email to check
 */
export function checkEmailExists(email: string): Action<CheckEmailExistsAction> {
    return buildAction(CHECK_EMAIL_ACTION, email);
}

/**
 * Build the Redux action to check if an email exists
 */
export function reset(): Action<null> {
    return buildAction(RESET_ACTION, null);
}

/** The shape of the response from the server on checking if an email address exists */
type CheckEmailExistsServiceResponse = {
    exists: boolean,
};

/**
 * Saga to check if an email address is already registered with the server
 * @param action the action to handle
 */
export function* checkEmailExistsSaga(action: Action<CheckEmailExistsAction>) : IterableIterator<any> {
    yield asyncAction(CHECK_EMAIL_ACTION, async (email: string) => {
        const response = await request<CheckEmailExistsServiceResponse>('/emails/{email}', {
            method: 'GET',
            urlParams: {
                email,
            }
        });

        return response.body.exists;
    }, action.payload);
}

/**
 * Reducer for when we start to check if an email address exists
 * @param state the state to update
 * @param action The action to handle
 */
export function CheckEmailExistsStartReducer(state: State, action: AsyncAction<string>) : State {
    return produce(state, (draft: State) => {
        draft.email = action.input;
        draft.state = CHECK_STATUS_START;
    });
}

/**
 * Reducer for when we start to check if an email address exists
 * @param state the state to update
 * @param action The action to handle
 */
export function CheckEmailExistsSuccessReducer(state: State, action: ResolvedAsyncAction<string, boolean>) : State {
    return produce(state, (draft: State) => {
        draft.state = action.payload ? CHECK_STATUS_EXISTS : CHECK_STATUS_UNKNOWN;
    });
}

/**
 * Reducer for when we reset the check if an email address exists
 * @param state the state to update
 */
export function ResetReducer(state: State) : State {
    return produce(state, (draft: State) => {
        delete draft.state;
        delete draft.email;
    });
}

/** Selector to get the state of the Check Email process */
export const selectCheckEmailStatus = buildSelector(['checkEmailExists'], (state: State) => state.state);

/** Selector to get the Email Address we are working with */
export const selectCheckEmailValue = buildSelector(['checkEmailExists'], (state: State) => state.email);

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [startedAction(CHECK_EMAIL_ACTION)] : CheckEmailExistsStartReducer,
    [succeededAction(CHECK_EMAIL_ACTION)] : CheckEmailExistsSuccessReducer,
    [RESET_ACTION] : ResetReducer
});

/** The sagas for this module */
export const sagas = [
    buildSaga(CHECK_EMAIL_ACTION, checkEmailExistsSaga),
];

/** The actual module */
export default {
    CHECK_STATUS_START,
    CHECK_STATUS_EXISTS,
    CHECK_STATUS_UNKNOWN,

    checkEmailExists,
    reset,

    selectCheckEmailStatus,
    selectCheckEmailValue
}
