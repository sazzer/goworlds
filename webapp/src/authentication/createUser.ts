import {createReducer} from "redux-create-reducer";
import {Action, asyncAction, buildAction, buildSaga} from "../redux";

/** Prefix for actions in this module */
const MODULE_PREFIX = 'CreateUser/';

/** The shape of the state for this module */
export interface State {
}

/** The initial state for this module */
const INITIAL_STATE: State = {
};

//////// Selectors for getting details from the state

//////// Action for creating a new user

/** The action type for creating a new user */
const CREATE_USER_ACTION = MODULE_PREFIX + 'createUser';

/** The shape of the action to create a new user */
declare type CreateUserAction = {
    email: string,
    name: string,
    password: string,
};

/**
 * Build the Redux action to create a new user
 * @param email the email to user
 * @param name The name to use
 * @param password The password to use
 */
export function createUser(email: string, name: string, password: string): Action<CreateUserAction> {
    return buildAction(CREATE_USER_ACTION, {
        email,
        name,
        password,
    });
}

/** The shape of the response from the server on creating a user */
type CreateUserServiceResponse = {
};

/**
 * Saga to actually create a new user
 * @param action the action to handle
 */
export function* createUserSaga(action: Action<CreateUserAction>) : IterableIterator<any> {
    yield asyncAction(CREATE_USER_ACTION, async (email: string) => {
        return action;
    }, action.payload);
}

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
});

/** The sagas for this module */
export const sagas = [
    buildSaga(CREATE_USER_ACTION, createUserSaga),
];

/** The actual module */
export default {
    createUser,
}
