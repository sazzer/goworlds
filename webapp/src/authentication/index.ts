import {createReducer} from "redux-create-reducer";
import produce from 'immer';
import {buildSaga, buildAction, Action, asyncAction} from "../redux";

/** The shape of the state for this module */
export type State = {

};

/** The initial state for this module */
const INITIAL_STATE: State = {

};

/** Prefix for actions in this module */
const MODULE_PREFIX = 'Authentication/';

//////// Check If Email Exists

const CHECK_EMAIL_ACTION = MODULE_PREFIX + 'checkEmailExists';

/** The shape of the action to check if an email exists */
type CheckEmailExistsAction = {
    email: string
};

/**
 * Build the Redux action to check if an email exists
 * @param email the email to check
 */
export function checkEmailExists(email: string): Action<CheckEmailExistsAction> {
    return buildAction(CHECK_EMAIL_ACTION, {
        email
    });
}

/**
 * Saga to check if an email address is already registered with the server
 * @param action the action to handle
 */
export function* checkEmailExistsSaga(action: Action<CheckEmailExistsAction>) : IterableIterator<any> {
    yield asyncAction(CHECK_EMAIL_ACTION, (email: string) => {
        return new Promise<string>((resolve) => {
            setTimeout(() => resolve('Success: ' + email), 5000);
        });
    }, action.payload.email);
}

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {

});

/** The sagas for this module */
export const sagas = [
    buildSaga(CHECK_EMAIL_ACTION, checkEmailExistsSaga),
];

/** The actual module */
export default {
    checkEmailExists
}
