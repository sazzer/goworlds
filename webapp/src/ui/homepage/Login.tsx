import React, {FunctionComponent} from 'react';
import {useDispatch} from 'react-redux'
import {Header, Segment} from "semantic-ui-react";
import {useTranslation} from "react-i18next";
import {EmailEntry} from "./EmailEntry";
import checkEmailExistsModule from "../../authentication/checkEmailExists";

/** The props that the Login area needs */
type LoginProps = {};

/**
 * The Login area
 * @constructor
 */
export const Login: FunctionComponent<LoginProps> = () => {
    const { t } = useTranslation();
    const dispatch = useDispatch();

    return (
        <Segment>
            <Header>
                {t('loginArea.header')}
            </Header>
            <EmailEntry onSubmit={(email) => dispatch(checkEmailExistsModule.checkEmailExists(email))}/>
        </Segment>
    );
};
