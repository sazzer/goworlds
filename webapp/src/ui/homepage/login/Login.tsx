import React, {FunctionComponent, useState} from 'react';
import {Button, Form} from "semantic-ui-react";
import {useTranslation} from "react-i18next";
import * as Yup from "yup";
import {Formik} from "formik";
import {FormikErrorMessage} from "../../common/FormikErrorMessage";

/** The props that the Login area needs */
type LoginProps = {
    email: string,
    onCancel: () => void,
};

/**
 * Component for logging in to an existing account
 * @constructor
 */
export const Login: FunctionComponent<LoginProps> = ({email, onCancel}) => {
    const { t } = useTranslation();
    const [submitting, setSubmitting] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);

    const schema = Yup.object().shape({
        email: Yup.string()
            .required(t('loginArea.email.errors.required')),
        password: Yup.string()
            .required(t('loginArea.password.errors.required')),
    });

    const doSubmit = (email: string, password: string) => {
        setSubmitting(true);
    };

    return (
        <Formik initialValues={{email: email, password: ''}}
                validationSchema={schema}
                onSubmit={(values) => doSubmit(values.email, values.password)}>
            {({values, isValid, errors, touched, handleSubmit, handleChange, handleBlur}) =>
                <Form onSubmit={handleSubmit} error={!isValid || error !== undefined} loading={submitting}>
                    <Form.Field required>
                        <label>
                            {t('loginArea.email.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('loginArea.email.placeholder')}
                                    name="email"
                                    type="text"
                                    value={values.email}
                                    readOnly />
                    </Form.Field>
                    <Form.Field required>
                        <label>
                            {t('loginArea.password.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('loginArea.password.placeholder')}
                                    name="password"
                                    type="password"
                                    value={values.password}
                                    error={touched.password && errors.password !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur} />
                        <FormikErrorMessage name="password" />
                    </Form.Field>
                    <Button type="submit" primary>
                        {t('loginArea.submit.register')}
                    </Button>
                    <Button type="reset" negative onClick={onCancel}>
                        {t('loginArea.submit.cancel')}
                    </Button>
                </Form>
            }
        </Formik>
    );
};
