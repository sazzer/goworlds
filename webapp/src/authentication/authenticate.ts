import {
    Action,
    asyncAction,
    buildAction,
    buildSaga,
    failedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request, RequestError} from "../api";
import {put} from "redux-saga-test-plan/matchers";
import {storeAccessToken} from "./accessToken";
import jwtDecode from 'jwt-decode';
import {storeCurrentUser} from "./currentUserId";
import {MODULE_PREFIX} from "./module";
import {getConfig} from "../config";

//////// Action for authenticating to request an Access Token and ID Token

/** The action type for checking if an email exists */
const AUTHENTICATE_ACTION = MODULE_PREFIX + 'authenticate';

/** The shape of the action to authenticate a user */
declare type AuthenticateAction = {
    email: string,
    password: string,
    callback: (err: Error | undefined) => void
}

/**
 * Build the Redux action to authenticate a user
 * @param email the email to authenticate with
 * @param password The password to authenticate with
 * @param callback A callback to trigger when the action is completed
 */
export function authenticate(email: string, password: string, callback?: (err: Error | undefined) => void): Action<AuthenticateAction> {
    return buildAction(AUTHENTICATE_ACTION, {
        email,
        password,
        callback: callback || (() => {})
    });
}

/** The shape of the response from the server on requesting an access token */
declare type AuthenticationServiceResponse = {
    access_token: string,
    expires_in: number,
    token_type: string,
    id_token: string,
};

/** An Error from an OAuth2 Request */
export class OAuth2Error extends Error {
    /** The OAuth2 error code */
    errorCode: string;

    /**
     * Construct the error
     * @param error the error code
     */
    constructor(error: string) {
        super(error);

        this.errorCode = error;
    }
}

/**
 * Saga to authenticate a user
 * @param action the action to handle
 */
export function* authenticateSaga(action: Action<AuthenticateAction>) : IterableIterator<any> {
    yield asyncAction(AUTHENTICATE_ACTION, async () => {
        const requestBody = new FormData();
        requestBody.set('grant_type', 'tag:goworlds,2019:oauth2/grant_type/email_password');
        requestBody.set('email', action.payload.email);
        requestBody.set('password', action.payload.password);
        requestBody.set('scope', 'openid');

        try {
            const response = await request<AuthenticationServiceResponse>('/oauth2/token', {
                method: 'POST',
                body: requestBody,
                clientId: process.env.REACT_APP_API_CLIENTID || getConfig('API_CLIENTID'),
            });

            return response.body;
        } catch (err) {
            if (err instanceof RequestError) {
                throw new OAuth2Error(err.body.error);
            }
            throw err;
        }
    }, action.payload);
}

/**
 * Saga to trigger when Authentication fails for some reason
 * @param action the action to handle
 */
export function authenticateFailedSaga(action: ResolvedAsyncAction<AuthenticateAction, Error>) {
    if (action.input) {
        action.input.callback(action.payload);
    }
}

/**
 * Saga to trigger when Authentication succeeded
 * @param action the action to handle
 */
export function* authenticateSucceededSaga(action: ResolvedAsyncAction<AuthenticateAction, AuthenticationServiceResponse>) : IterableIterator<any> {
    if (action.payload) {
        const expiry = new Date(new Date().getTime() + (action.payload.expires_in * 1000));
        yield put(storeAccessToken(action.payload.access_token, expiry));

        const idToken = jwtDecode(action.payload.id_token) as any;
        if (idToken.sub) {
            yield put(storeCurrentUser(idToken.sub));
        }
    }

    if (action.input) {
        action.input.callback(undefined);
    }
}

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(AUTHENTICATE_ACTION, authenticateSaga),
    buildSaga(succeededAction(AUTHENTICATE_ACTION), authenticateSucceededSaga),
    buildSaga(failedAction(AUTHENTICATE_ACTION), authenticateFailedSaga),
];
