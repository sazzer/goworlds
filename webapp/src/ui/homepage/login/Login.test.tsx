import configureStore from "redux-mock-store";
import {render, fireEvent} from '@testing-library/react';
import {Provider} from "react-redux";
import React from "react";
import {Login} from "./Login";
import {OAuth2Error} from "../../../authentication/authenticate";

function setup({email} = {email: 'graham@grahamcox.co.uk'}) {
    const mockStoreCreator = configureStore();
    const store = mockStoreCreator({});

    const onCancelCallback = jest.fn();

    const element = render(<Provider store={store}><Login email={email} onCancel={onCancelCallback}/></Provider>);

    const enterPassword = (password: string) => {
        fireEvent.change(element.getByPlaceholderText('Enter Password'), {
            target: {
                value: password,
            }
        });
    };

    const submitForm = () => new Promise(resolve => {
        fireEvent.click(element.getByText('Register'));
        setImmediate(resolve);
    });

    const cancelForm = () => fireEvent.click(element.getByText('Cancel'));

    return {
        onCancelCallback,
        store,

        element,

        enterPassword,
        submitForm,
        cancelForm
    }
}

it('Initially renders correctly', () => {
    const {element} = setup();

    expect(element.baseElement).toMatchSnapshot();
});

it('Triggers the Cancel Callback when the Cancel button is pressed', () => {
    const {element, cancelForm, onCancelCallback} = setup();

    cancelForm();

    expect(onCancelCallback).toBeCalledTimes(1);
});

it('Renders correctly after entering a password', () => {
    const {element, enterPassword} = setup();

    enterPassword('password');

    expect(element.baseElement).toMatchSnapshot();
});

it('Renders an error if submitted without a password', async () => {
    const {element, submitForm} = setup();

    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Renders as loading if submitted with a password', async () => {
    const {element, enterPassword, submitForm} = setup();

    enterPassword('password');
    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Triggers the correct Redux Action when submitted with a password', async () => {
    const {enterPassword, submitForm, store} = setup();

    enterPassword('password');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()[0]).toMatchObject({
        type: 'Authentication/authenticate',
        payload: {
            email: 'graham@grahamcox.co.uk',
            password: 'password'
        }
    });
});

it('Updates correctly when the callback indicates success', async () => {
    const {element, enterPassword, submitForm, store} = setup();

    enterPassword('password');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(undefined);

    expect(element.baseElement).toMatchSnapshot();
});

it('Updates correctly when the callback indicates an unexpected error', async () => {
    const {element, enterPassword, submitForm, store} = setup();

    enterPassword('password');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(new Error());

    expect(element.baseElement).toMatchSnapshot();
});

it('Updates correctly when the callback indicates an access_denied error', async () => {
    const {element, enterPassword, submitForm, store} = setup();

    enterPassword('password');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(new OAuth2Error('access_denied'));

    expect(element.baseElement).toMatchSnapshot();
});
