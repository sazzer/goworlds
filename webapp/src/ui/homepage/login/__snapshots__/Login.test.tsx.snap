// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initially renders correctly 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Renders an error if submitted without a password 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="error field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="ui error message"
        >
          Please enter a password
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Renders as loading if submitted with a password 1`] = `
<body>
  <div>
    <form
      class="ui loading form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="password"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Renders correctly after entering a password 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="password"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Updates correctly when the callback indicates an access_denied error 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="password"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      >
        Invalid password or account blocked
      </div>
    </form>
  </div>
</body>
`;

exports[`Updates correctly when the callback indicates an unexpected error 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="password"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      >
        An unexpected error occurred
      </div>
    </form>
  </div>
</body>
`;

exports[`Updates correctly when the callback indicates success 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="Login"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              readonly=""
              type="text"
              value="graham@grahamcox.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="password"
      >
        <label>
          Password
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="password"
              placeholder="Enter Password"
              type="password"
              value="password"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Register
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Cancel
      </button>
      <div
        class="ui error message"
        data-test="LoginErrors"
      />
    </form>
  </div>
</body>
`;
